src/object_async/hello_async.cpp

```diff
@@ -0,0 +1,227 @@
+#include "hello_async.hpp"
+
+#include <exception>
+#include <stdexcept>
+#include <iostream>
+#include <map>
+
+// If this was not defined within a namespace, it would be in the global scope.
+namespace object_async {
+
+  // Custom constructor, assigns custom name passed in from Javascript world.
+  // This constructor uses member init list via the semicolon, aka "direct initialization" 
+  // which is more efficient than using assignment operators.
+  HelloObjectAsync::HelloObjectAsync(std::string name) : 
+    name_(name) {}
+
+  // Triggered from Javascript world when calling "new HelloObjectAsync()" or "new HelloObjectAsync(name)"
+  NAN_METHOD(HelloObjectAsync::New) {
+    if (info.IsConstructCall())
+    {
+        try
+        {   
+            if (info.Length() >= 1) {
+              if (info[0]->IsString()) 
+              {
+                std::string name = *v8::String::Utf8Value(info[0]->ToString());
+                auto *const self = new HelloObjectAsync(name);
+                self->Wrap(info.This());
+              }
+              else
+              {
+                return Nan::ThrowTypeError(
+                    "arg must be a string");
+              }
+            }
+            else {
+                auto *const self = new HelloObjectAsync();
+                self->Wrap(info.This());
+            } 
+
+        }
+        catch (const std::exception &ex)
+        {
+            return Nan::ThrowTypeError(ex.what());
+        }
+
+        info.GetReturnValue().Set(info.This());
+    }
+    else
+    {
+        return Nan::ThrowTypeError(
+            "Cannot call constructor as function, you need to use 'new' keyword");
+    }
+  }
+
+  /*
+  * This is an internal function used to return callback error messages instead of
+  * throwing errors.
+  * Usage:
+  *
+  * v8::Local<v8::Function> callback;
+  * CallbackError("error message", callback);
+  * return; // this is important to prevent duplicate callbacks from being fired!
+  */
+  inline void CallbackError(std::string message, v8::Local<v8::Function> callback) {
+  	v8::Local<v8::Value> argv[1] = { Nan::Error(message.c_str()) };
+  	Nan::MakeCallback(Nan::GetCurrentContext()->Global(), callback, 1, argv);
+  }
+
```

Comment:
> @springmeyer so inline is great for reusable methods within a file?

- inlining, when in a cpp file, may help performance.
- inlining, when in an hpp file, may help performance and may also be absolutely necessary to allow the header to be included from multiple cpp files when the code in the header includes both the definition and the implementation (which a "header-only" library does). And example of this being fixed is https://github.com/mapbox/vector-tile/pull/19.

So, inlining is important to understand because it impacts performance, code size, and is critical to header-only library design.

> And a util header is one step further than that, great for reusable methods within an application/directory

Exactly. If you want to share code in multiple .cpp files then you need to put at least definitions for it in an .hpp file (and, in the case of header-only design, also the implementation)

Ticketed adding this to the C++ glossary at https://github.com/mapbox/cpp/issues/28